### PR TITLE
[es] Add remote read clusters option for cross-cluster querying

### LINKED ISSUE #1883: Elasticsearch Cross Cluster Query
Would like to add the ability to use the elasticsearch cross cluster query api to query spans across multiple AZs.

I already have some general idea on how to do it but wanted to check here before spending the time working on an MR.

Initial design:

`--es.remote-read-clusters=local,cluster_one,cluster_two`

This will result in the search query GET /jaeger-span-2019-10-28,cluster_one:jaeger-span-2019-10-28,cluster_two:jaeger-span-2019-10-28/_search

es/options.go

```go
flagSet.String(
		nsConfig.namespace+suffixRemoteReadClusters,
		nsConfig.remoteReadClusters,
		"(experimental) The comma-separated list of Elasticsearch remote clusters to read from. Add local to the list to also target the main cluster. i.e. local,cluster1,cluster2")
```

es/spanstore/reader.go

Used a wrapper function to inject the remote clusters. This is the best way I could come up with that has the minimal affect on current code.

```go
func getTimeRangeIndexFn(archive, useReadWriteAliases bool, remoteReadClusters string[]) timeRangeIndexFn {
	if archive {
		var archivePrefix string
		if useReadWriteAliases {
			archivePrefix = archiveReadIndexSuffix
		} else {
			archivePrefix = archiveIndexSuffix
		}
		return addRemoteReadClusters(func(indexName string, startTime time.Time, endTime time.Time) []string {
			return []string{archiveIndex(indexName, archivePrefix)}
		}, remoteReadClusters)
	}

	if useReadWriteAliases {
		return addRemoteReadClusters(func(indices string, startTime time.Time, endTime time.Time) []string {
			return []string{indices + "read"}
		}, remoteReadClusters)
	}

	return addRemoteReadClusters(timeRangeIndices, remoteReadClusters)
}

// Elasticsearch cross cluster query api example: GET /twitter,cluster_one:twitter,cluster_two:twitter/_search
// Add remote cluster prefix to each index
func addRemoteReadClusters(fn timeRangeIndexFn, remoteReadClusters string[]) timeRangeIndexFn {
	return func(indices string, startTime time.Time, endTime time.Time) []string {
		jaegerIndices := fn(indices, startTime, endTime)
		if len(remoteReadClusters) > 0 {
			remoteJaegerIndices := string[]{}
			for remoteCluster := range remoteReadClusters {
				for jaegerIndex := range jaegerIndices {

					if remoteCluster != "local"{
						jaegerIndex = remoteCluster + ":" + jaegerIndex
					}

					remoteJaegerIndices = append(remoteJaegerIndices, jaegerIndex)
				}
			}
		} else {
			return jaegerIndices
		}
	}
}
```

### PR DESCRIPTION
**Pull request by dgrizzanti**
## Which problem is this PR solving?
- Enables elasticsearch [cross-cluster querying](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-cross-cluster-search.html)

Resolves #1883
Updates to PR #1892

## Short description of the changes
The cross cluster query api functions by appending a pre-configured remote cluster to an index e.g. cluster_one:index_name. This change simply applies a prefix for each remote cluster and for each index. You will end up with `total_indicies = indicies * remote_clusters`. Also added a special case local which will not append a prefix and query the local cluster.

This works with both standard and rollover indices since it mostly comes down to a simple string manipulation.

### CODE REVIEW

File: plugin/storage/es/spanstore/reader_test.go [hunk 1]
```diff
@@ -149,35 +149,40 @@ func TestSpanReaderIndices(t *testing.T) {
...
 		{params: SpanReaderParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
 			IndexPrefix: "foo:", Archive: true},
-			index: "foo:" + indexPrefixSeparator + spanIndex + archiveIndexSuffix},
+			indices: []string{"foo:" + indexPrefixSeparator + spanIndex + archiveIndexSuffix}},
 		{params: SpanReaderParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
 			IndexPrefix: "foo:", Archive: true, UseReadWriteAliases: true},
-			index: "foo:" + indexPrefixSeparator + spanIndex + archiveReadIndexSuffix},
+			indices: []string{"foo:" + indexPrefixSeparator + spanIndex + archiveReadIndexSuffix}},
+		{params: SpanReaderParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
+			IndexPrefix: "", Archive: false, RemoteReadClusters: []string{"cluster_one", "cluster_two"}},
```
Comment (albertteoh) [inline plugin/storage/es/spanstore/reader_test.go:177]: It'd be good to add two more test cases for `RemoteReadClusters: []string{"cluster_one", "cluster_two"}}` together with because the `addRemoteReadClusters` calls could be removed for archive and alias mode without test failure:
- `UseReadWriteAliases: true`
- `Archive: true`

File: plugin/storage/es/spanstore/reader_test.go [hunk 2]
```diff
@@ -149,35 +149,40 @@ func TestSpanReaderIndices(t *testing.T) {
...
+			indices: []string{spanIndex + archiveIndexSuffix}},
 		{params: SpanReaderParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
 			IndexPrefix: "foo:", Archive: true},
-			index: "foo:" + indexPrefixSeparator + spanIndex + archiveIndexSuffix},
+			indices: []string{"foo:" + indexPrefixSeparator + spanIndex + archiveIndexSuffix}},
 		{params: SpanReaderParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
 			IndexPrefix: "foo:", Archive: true, UseReadWriteAliases: true},
-			index: "foo:" + indexPrefixSeparator + spanIndex + archiveReadIndexSuffix},
+			indices: []string{"foo:" + indexPrefixSeparator + spanIndex + archiveReadIndexSuffix}},
+		{params: SpanReaderParams{Client: client, Logger: logger, MetricsFactory: metricsFactory,
```
Comment (albertteoh) [inline plugin/storage/es/spanstore/reader_test.go:176]: Please refer to @yurishkuro's [comment](https://github.com/jaegertracing/jaeger/pull/1892/files#r344831587) about making the test cases multiline.

File: plugin/storage/es/spanstore/reader.go [hunk 3]
```diff
@@ -139,24 +140,43 @@ type timeRangeIndexFn func(indexName string, indexDateLayout string, startTime t
...
-		return func(indices string, indexDateLayout string, startTime time.Time, endTime time.Time) []string {
+		return addRemoteReadClusters(func(indices string, indexDateLayout string, startTime time.Time, endTime time.Time) []string {
 			return []string{indices + "read"}
+		}, remoteReadClusters)
+	}
+	return addRemoteReadClusters(timeRangeIndices, remoteReadClusters)
+}
+
+// Elasticsearch cross cluster query api example: GET /twitter,cluster_one:twitter,cluster_two:twitter/_search
+// Add a remote cluster prefix for each cluster and for each index and add it to the list of original indices
```
Comment (albertteoh) [inline plugin/storage/es/spanstore/reader.go:164]: nit: Please end comment sentences with a full-stop. 

Same as [above](https://github.com/jaegertracing/jaeger/pull/2874/files#diff-de5aae4da3cb03d3a73919aca1f6f38181b906dfd205bb0def4bf7b9d6042992R254) though I understand this convention is not followed in the adjacent functions below it.

**Comment by dgrizzanti** [comment]
@albertteoh thanks for the comments and suggestions. I think I covered all of them if you can take a look again. Thanks!

File: plugin/storage/integration/integration.go [hunk 4]
```diff
@@ -36,7 +36,7 @@ import (
 )
 
 const (
-	iterations = 30
+	iterations = 100
```
Comment (albertteoh) [inline plugin/storage/integration/integration.go:39]: Were these changes brought in from #2873?

**Comment by albertteoh** [comment]
> @albertteoh thanks for the comments and suggestions. I think I covered all of them if you can take a look again. Thanks!

[One more](https://github.com/jaegertracing/jaeger/pull/2874/files#r592908331) :) Also, it seems another commit was pulled into this PR: https://github.com/jaegertracing/jaeger/pull/2874/commits/510523707eec305870eafa17845459aaf7bd7a04.

File: plugin/storage/integration/integration.go (hunk 4 above)
Comment (dgrizzanti) [reply plugin/storage/integration/integration.go:39]: I did rebase so may be they got pulled in by accident? I flip it back so it doesn't show up in this PR.

File: plugin/storage/es/spanstore/reader.go [hunk 5]
```diff
@@ -139,24 +140,44 @@ type timeRangeIndexFn func(indexName string, indexDateLayout string, startTime t
...
+// Elasticsearch cross cluster api example GET /twitter,cluster_one:twitter,cluster_two:twitter/_search.
+func addRemoteReadClusters(fn timeRangeIndexFn, remoteReadClusters []string) timeRangeIndexFn {
+	return func(indexPrefix string, indexDateLayout string, startTime time.Time, endTime time.Time) []string {
+		jaegerIndices := fn(indexPrefix, indexDateLayout, startTime, endTime)
+		if len(remoteReadClusters) == 0 {
+			return jaegerIndices
 		}
+
+		for _, jaegerIndex := range jaegerIndices {
+			for _, remoteCluster := range remoteReadClusters {
```
Comment (albertteoh) [inline plugin/storage/es/spanstore/reader.go:173]: Related to the earlier comment, if `remoteReadClusters` defaults to `[]string{""}`, the returned indices count would be doubled with the new indices looking like `:<index-name>`. Is that expected?

File: plugin/storage/integration/integration.go (hunk 4 above)
Comment (albertteoh) [reply plugin/storage/integration/integration.go]: Yup, I had more luck with `git merge` and did recall having similar problems when using `git rebase`. Thanks 🙏

File: pkg/es/config/config.go [hunk 6]
```diff
@@ -193,6 +195,9 @@ func (c *Configuration) NewClient(logger *zap.Logger, metricsFactory metrics.Fac
 
 // ApplyDefaults copies settings from source unless its own value is non-zero.
 func (c *Configuration) ApplyDefaults(source *Configuration) {
+	if len(c.RemoteReadClusters) == 0 {
+		c.RemoteReadClusters = source.RemoteReadClusters
```
[the remark this answers is not in the document]
Comment (dgrizzanti) [reply pkg/es/config/config.go:199]: Thanks @albertteoh - I didn't realize this nuance with `strings.Split()` but after looking at the docs, this makes sense. I didn't see a great (easy) way around this since the cli takes in strings, so I updated `options.go` to check the length and only split when necessary (otherwise default to `[]string{}`). I added 2 tests cases around options as well to cover this.

File: plugin/storage/es/spanstore/reader.go (hunk 5 above)
Comment (dgrizzanti) [reply plugin/storage/es/spanstore/reader.go:173]: This should be avoided now with the changes to `plugin/storage/es/options.go`

File: plugin/storage/es/options_test.go [hunk 7]
```diff
@@ -103,6 +106,19 @@ func TestOptionsWithFlags(t *testing.T) {
...
+	opts := NewOptions("es", "es.aux")
+	v, command := config.Viperize(opts.AddFlags)
+	err := command.ParseFlags([]string{
+		"--es.remote-read-clusters=",
+	})
+	require.NoError(t, err)
+	opts.InitFromViper(v)
+
+	primary := opts.GetPrimary()
+	assert.Equal(t, []string{}, primary.RemoteReadClusters)
```
Comment (albertteoh) [inline plugin/storage/es/options_test.go:119]: I think the following would be more expressive/readable: `assert.Empty(t, primary.RemoteReadClusters)`

File: pkg/es/config/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (albertteoh) [reply pkg/es/config/config.go:199]: Thanks for adding the test cases to cover the empty string. I think there's also the following case: `--es.remote-read-clusters=cluster_one,,cluster_two`, where the second item is empty but I think we can consider this a user error.

**Review by albertteoh — APPROVED** [review]
lgtm, thanks @dgrizzanti.

**Comment by yurishkuro** [comment]
Thank you!
